torch/nested/_internal/nested_tensor.py

```diff
@@ -568,84 +555,181 @@ def jagged_from_tensor_and_lengths(
+) -> Tuple[CachedTensor, Optional[torch.Tensor]]:
+    # 1. Constructs a fresh CachedTensor from provided metadata
+    # - normalizes all fields
+    #   - unwraps CachedTensor
+    #   - puts lengths/offsets on correct field based on device
+    #   - stuffs min/max seqlen into tensor if necessary
+    # - if a piece of metadata is present multiple times, we arbitrarily
+    #   choose one
+    # 2. Marks dynamic for max/min seqlen if present
+    # 3. Optionally returns non_contig_offsets if lengths is present
+    #    This means that it is not possible to construct a contiguous
+    #    NestedTensor with lengths metadata today.
+    assert offsets is not None
+
+    metadata: Dict[str, Optional[torch.Tensor]] = dict.fromkeys(
+        SOURCE_FIELDS + EXTRA_FIELDS
+    )
+
+    def process_cached_tensor(metadata: Dict, t: torch.Tensor):
+        assert isinstance(t, CachedTensor)
+        for k, v in t.metadata.items():
+            if v is not None:
+                metadata[k] = v
 
-    max_seqlen_tensor = None
+    def process_raw_source_tensor(metadata: Dict, key: str, t: torch.Tensor):
+        if t.is_cpu:
+            metadata[f"_host_{key}"] = t
+        else:
+            metadata[f"_device_{key}"] = t
+
+    def process_min_max_seqlen(
+        metadata: Dict, key: str, o: Union[torch.Tensor, int, torch.SymInt]
+    ):
+        if isinstance(o, (int, torch.SymInt)):
+            metadata[f"_{key}_tensor"] = _store_val_in_tensor(o)
+        else:
+            metadata[f"_{key}_tensor"] = o
+
+    if offsets is not None and lengths is not None:
+        non_contig_offsets = offsets
+        if isinstance(lengths, CachedTensor):
+            process_cached_tensor(metadata, lengths)
+        else:
+            process_raw_source_tensor(metadata, "lengths", lengths)
+    else:
+        non_contig_offsets = None
+        if isinstance(offsets, CachedTensor):
+            process_cached_tensor(metadata, offsets)
+        else:
+            process_raw_source_tensor(metadata, "offsets", offsets)
+
+    if min_seqlen is not None:
+        process_min_max_seqlen(metadata, "min_seqlen", min_seqlen)
     if max_seqlen is not None:
-        max_seqlen_tensor = _store_val_in_tensor(max_seqlen)
+        process_min_max_seqlen(metadata, "max_seqlen", max_seqlen)
+
+    if metadata.get("_max_seqlen_tensor") is not None:
+        torch._dynamo.mark_dynamic(metadata["_max_seqlen_tensor"], 0)
+    if metadata.get("_min_seqlen_tensor") is not None:
+        torch._dynamo.mark_dynamic(metadata["_min_seqlen_tensor"], 0)
+
+    metadata_tensor = make_cached_tensor_for_nested(metadata)
+    return metadata_tensor, non_contig_offsets
+
+
+# Useful for tests that previously constructed the NestedTensor directly via
+# the raw constructor.
+# kwargs that are not _metadata_cache or _ragged_idx are ignored.
+# Fields in _metadata_cache that are not _metadata_cache are ignored.
+def _construct_nested_tensor_compat(
```

Comment:
I have a strong preference for defining this in the test code instead of here. I want to minimize the number of APIs people can choose to use (even if they shouldn't) that make our lives harder wrt BC.

Reply:
It's used in ViewBufferFromNested, which... we can probably remove, but I I'm also not sure I want to do in this PR. 
I totally agree with your point though - would it help if I rename it _DO_NOT_USE_construct_nested_tensor_compat temporarily until we remove ViewBufferFromNested.